sum_test.go

```diff
@@ -122,75 +123,22 @@ func BenchmarkBlake2B(b *testing.B) {
 	}
 }
 
-// Test that the identity hash function checks
-// its `length` arguemnt for values that are
-// different to its `data` length.
-func TestSmallerLengthHashID(t *testing.T) {
```

Comment:
I made a comment at the top about this.

It appeared to be testing an API that wasn't actually exported... and the nearest API that is exported (Sum) has a general contract of truncating a hash upon short length, so it was overall unclear what this test should be checking.

So, I'm game to put this, or something like it, back in.  But it's just not clear to me what it's supposed to be doing -- if I port it in the most literal possible way (e.g. to call `Sum`), it fails, but "fixing" it would change the general contract of `Sum` -- so I'd want someone to chime in on what the expected behavior should actually be.

Reply:
It ensures that ID hashes _cannot_ truncate (nonsensical). We passed the hash length to the hasher specifically to allow this check.

I guess we have two easy solutions:

1. Make the hasher constructor take a size. We'd probably need two register functions to handle this (`Register` and `RegisterSized`, or something like that).
2. Add an optional `MinSize() int` function to our hashers (i.e., type assert to see if this is defined). When truncating, check to see if there is a `MinSize()` defined and, if there is, make sure the requested size is at least the minimum size.
3. Allow _registering_ a minimum size for a codec.

2 or 3 probably make the most sense where 3 is likely the most performant.

Reply:
Okay, I'm going to revive this test and push a change to the `Sum` function which explicitly rejects truncations for identity multihashes.  That's the most conservative change that gets this test back and gets it passing and maintains the old behavior and its limits.

The MinSize idea sounds interesting (especially in light of https://github.com/ipfs/go-verifcid/ , which iirc is also concerned about overly short hashes) but sounds like it'll broach a bigger discussion than I wanna hitch to this PR.